sql/src/test/java/io/crate/operation/scalar/AbstractScalarFunctionsTest.java

```diff
@@ -219,4 +219,66 @@ public Object value() {
             throw new AssertionError("Input.value() should only be called once");
         }
     }
+
+    private static class InputApplierContext implements Iterator<Input> {
+
+        private final Iterator<Input> inputsIterator;
+        private final SqlExpressions sqlExpressions;
+        private final boolean normalize;
+
+        InputApplierContext(Input[] inputs, SqlExpressions sqlExpressions, boolean normalize) {
+            this.inputsIterator = Arrays.asList(inputs).iterator();
+            this.sqlExpressions = sqlExpressions;
+            this.normalize = normalize;
+        }
+
+        public Input next() {
+            return inputsIterator.next();
+        }
+
+        @Override
+        public boolean hasNext() {
+            return inputsIterator.hasNext();
+        }
+
+        @Override
+        public void remove() {
+        }
+    }
+
+    private static class InputApplier extends SymbolVisitor<InputApplierContext, Input> {
+
+        @Override
+        public Input visitLiteral(Literal symbol, InputApplierContext context) {
+            return symbol;
+        }
+
+        @Override
+        public Input visitField(Field field, InputApplierContext context) {
+            if (!context.hasNext()) {
+                return null;
+            }
+            return context.next();
+        }
+
+        @Override
+        public Input visitFunction(Function function, InputApplierContext context) {
+            Input[] argInputs = new Input[function.arguments().size()];
+            for (int j = 0; j < function.arguments().size(); j++) {
+                Input input = function.arguments().get(j).accept(this, context);
+                argInputs[j] = input;
+                // replace arguments on function for normalization
+                if (context.normalize && input instanceof Literal) {
+                    function.arguments().set(j, (Literal) argInputs[j]);
+                }
+            }
+
+            if (context.normalize) {
+                return (Input) context.sqlExpressions.normalize(function);
+            }
+
+            Scalar scalar = (Scalar) TestingHelpers.getFunctions().get(function.info().ident());
```

Comment:
I think it would be cheaper to call sqlExpressions.functions() - the testingHelpers method will re-create the Functions class each time.